README.md
# craggy2019

---

**Table of Contents**

1. [cRaggy is Back - Looking at Housing and Evictions in King County](#craggy-is-back---looking-at-housing-and-evictions-in-king-county)
2. [Data Source](#data-source)
3. [Accessing the data](#accessing-the-data)

    3.1 [Add your code to the existing source code](#add-your-code-to-the-existing-source-code)  
    3.2 [Use the `craggy2019` package from your own independent project](#use-the-craggy2019-package-from-your-own-independent-project)

4. [Examples and kick-starters](#examples-and-kick-starters)
5. [Notes and resources](#notes-and-resources)
6. [License](#license)

---

<!-- badges: start -->
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- badges: end -->

> Disclaimer: this package is experimental and intended for use simply as an informal exercise for working with data in the [R language](https://www.r-project.org/). Read the [NEWS.md](NEWS.md) for recent updates.

---

## cRaggy is Back - Looking at Housing and Evictions in King County

cRaggy, our data exploration and visualization activity is back! For those of you unfamiliar with cRaggy, the [Portland R User Group](https://www.meetup.com/portland-r-user-group/) makes a large dataset available for visualization and analysis, so that you can present and share your results, visualizations, and insights about the dataset.

We will be looking at a combination of housing data and eviction data for King County, Washington, United States, hoping to gain insights about the nature of evictions in King County. This dataset comes courtesy of the planners of the [2019 Symposium for Data Science and Statistics](https://ww2.amstat.org/meetings/sdss/2019/). 

We encourage you to collaborate together and work on visualizations! Our goal is to encourage collaboration and share our code and insights.

In September, we'll have a visualization gallery and lightning talks about the visualizations and insights you've come up with. Event information will be posted on our [Meetup page](https://www.meetup.com/portland-r-user-group/events/).

All are welcome. Participation in cRaggy requires agreeing to the [Code of Conduct](.github/CODE_OF_CONDUCT.md)

## Data Source

King County Eviction Data for cRaggy 2019.

This dataset came from the [SDSS 2019 Data Hack Activity](https://ww2.amstat.org/meetings/sdss/2019/).

## Accessing the data

This project is organized as an [R package](https://r-pkgs.org/), which
contains the data and includes some examples of reading, exploring, analyzing, and visualizing the data. There are several ways to use this project.

### Add your code to the existing source code

**Context**: You want to use the existing sample code and write your own code within the same source structure of this package. You might want to contribute some of your code back to this project, but you are not sure about that yet.

**Solution**: [Fork](https://help.github.com/en/articles/fork-a-repo) the project, so that you can work in your own version. A "fork" is linked back to the original code, so that you can contribute changes later, if you so choose, by making a [Pull Request](https://help.github.com/en/articles/creating-a-pull-request-from-a-fork).

1. [Fork](https://help.github.com/en/articles/fork-a-repo) this project, which creates a separate copy in _your_ Github account.
1. [Clone](https://help.github.com/en/articles/cloning-a-repository) the repository that is now forked in _your_ Github account to your local machine.
1. Open the copy on your local machine in [RStudio](https://www.rstudio.com/products/rstudio/).
1. Click on the `Build > Install and Restart` button, which in the default application configuration is in the `Environment, History, ...` pane in the upper right of the application window.
1. The package will be built, installed in your local package library, and loaded into your session. When you see `library(craggy2019)` executed in the console, then the build and load is complete and your environment is ready for you to use the package.
1. Now you can use the [base::system.file](https://www.rdocumentation.org/packages/base/versions/3.6.1/topics/system.file) function to reference a data file within the loaded package. For example,
    ```r
    readr::read_csv(system.file("extdata", "evictions.csv", package = "craggy2019"))
    ```
    This will read data from the loaded package in the R Session, _not_ from the file system.
1. Alternatively, you can refer to the data files through the file system in the `inst/extdata` directory. For example,
    ```r
    here::here("inst", "extdata", "evictions.csv")`.
    ```
1. As you write code and perhaps add more data, you will [commit](https://git-scm.com/docs/git-commit) those changes with `git` on your local machine. From time to time, you should [push](https://help.github.com/en/articles/pushing-commits-to-a-remote-repository) those changes to _your_ Github account. Since you [forked](https://help.github.com/en/articles/fork-a-repo) the project, those changes will only affect _your_ repository, not the [pdxlang/craggy_2019](https://github.com/pdxrlang/craggy_2019) repository from which you forked.
1. As mentioned previously, you can make a [Pull Request](https://help.github.com/en/articles/creating-a-pull-request-from-a-fork) from your forked repository to the original [pdxlang/craggy_2019](https://github.com/pdxrlang/craggy_2019) repository, in order to have your code merged into the original source. However, we do not explain how to do that here.

### Use the `craggy2019` package from your own independent project

**Context**: You already have a project where you want to write code to read, explore, analyze, and visualize the Evictions data. You do _not_ want to write your code within the framework of this `craggy2019` package source code.

**Solution**: You need to install and load this `craggy2019` package in your local environment, so that you can access the data that it contains. However, you do _not_ need to [fork](https://help.github.com/en/articles/fork-a-repo) or [clone](https://git-scm.com/docs/git-clone) the [pdxlang/craggy_2019](https://github.com/pdxrlang/craggy_2019) repository.

1. Install the `craggy2019` package:
    ```r
        install.packages("remotes")
        library(remotes)
        remotes::install_github("pdxrlang/craggy_2019")
        # Note: Unfortunately, the repository name and package name do not match!
        library(craggy2019)
    ```
1. Now you can use the [base::system.file](https://www.rdocumentation.org/packages/base/versions/3.6.1/topics/system.file) function to reference a data file within the package. For example,
    ```r
    readr::read_csv(system.file("extdata", "evictions.csv", package = "craggy2019"))
    ```
    This will read data from the loaded package in the R Session, _not_ from the file system.
1. FYI: With [system.file](https://www.rdocumentation.org/packages/base/versions/3.6.1/topics/system.file), you do not reference the `inst` directory level, which is the root level for the data files in the _source code_, but it is _not_ the root level for the data in the _loaded package_. When a package is loaded, everything within the [`inst` directory](http://r-pkgs.had.co.nz/inst.html) is placed at the root level of the package. So you start at the `extdata` directory level, which is the level that holds all of the data files that we are using for this project.
1. FYI: `inst/extdata` is the R package convention for [where data is placed within a package](https://r-pkgs.org/data.html). Almost all packages follow that convention, so we followed it for this project too.

## Examples and kick-starters

These examples can help you start analyzing the data:

  * Hosted [burro](https://laderast.github.io/burro/) demo: [Browse King County evictions data](https://tladeras.shinyapps.io/evictions_king_county/).
  * [Run burro in your local environment](vignettes/evictions.Rmd).
  * [Introduction to Geospatial Visualization](vignettes/Introduction-to-Visualizing-Geospatial-Data.md) for the King County data.
  * [Read data from the `oneNightCount.csv` file](vignettes/demo-1night-vroom-etc.md) with [vroom](https://cran.r-project.org/package=vroom) and summarize the contents with [skimr](https://cran.r-project.org/package=skimr) and [inspectdf](https://cran.r-project.org/package=inspectdf).

## Notes and resources

See the [NEWS](NEWS.md) for recent code changes, as well as links to resources outside of this project.

## License

This code is licensed under the [MIT License](LICENSE.md).


NEWS.md
# craggy2019 (development version)

* Converted from a simple R project to an [R package](http://r-pkgs.had.co.nz/package.html). Unfortunately, the project file needed to be renamed from `craggy_2019` to `craggy2019`, in order to conform to the [requirements for a package name](http://r-pkgs.had.co.nz/package.html#naming). However, developing the project as a package has several advantages:

    * [RStudio](https://www.rstudio.com/products/rstudio/) and 
        the [devtools](https://cran.r-project.org/package=devtools) package 
        provide a lot of tooling that makes it easier to work with the 
        data and code:

        * `.Rmd` files in the `vignettes` directory are run automatically during the build process to produce `html` and `md` files.
        * There is an explicit listing of package dependencies in the `DESCRIPTION` file.
        * Unit tests can be run automatically with `devtools::test()`.
        * Package and function documentation are generated automatically during the build process and are accessible from the console, e.g. `?'craggy2019-package'`.
        * `devtools::check()` does a lot of validation of code.

    * The directory structure for data, code, and tests follows the standard conventions used by all R packages. That makes it easy for an experienced R developer to browse the project quickly.
    * Packages simplify the process for developing reusable code that is accessible within this package or from other projects. Rather than copying and pasting between `.Rmd` files, cRaggy participants can extract reusable code into functions defined in the package, which can then be re-used in multiple `.Rmd` files.
    * The package structure also encourages the development of unit tests for code, because it is easy to write unit tests for code that has been isolated in a function definition.

* [aggRegate Meeting Notes 2019-08-20 & 22](https://docs.google.com/document/d/1wRCDgZkA1fUBX2JonwGENTlkEP5fl6tdF_eDeNbJzVc/edit)

* [burro](https://laderast.github.io/burro/) application to [browse Evictions data](https://tladeras.shinyapps.io/evictions_king_county/).

* [aggRegate Meeting Notes 2019-07-18](https://docs.google.com/document/d/165pu2Sm4OgeMZx-tBKsfLX6FFqbbrw_c2m7lA1S9iKg/edit)


tests/testthat.R
library(testthat)
library(craggy2019)

test_check("craggy2019")


tests/testthat/test_get_king_pit.R
library(testthat)
TOLERANCE = 0.0001
EXPECTED_MAX_YEAR <- 2018
EXPECTED_MIN_YEAR <- 2007

context("getting PIT for King County")

test_that("data set has the expected years", {
  pit <- get_king_pit()

  expect_equal(max(pit$year), EXPECTED_MAX_YEAR)
  expect_equal(min(pit$year), EXPECTED_MIN_YEAR)

  actual_n_rows = dim(pit)[1]
  expect_equal(actual_n_rows, EXPECTED_MAX_YEAR - EXPECTED_MIN_YEAR + 1)
})



R/get_king_pit.R
library(dplyr)
library(rio)

# globalVariables hack that avoids "Notes" from being issued
# by devtools::check().
# See:
# https://stackoverflow.com/questions/9439256/how-can-i-handle-r-cmd-check-no-visible-binding-for-global-variable-notes-when
# for the solution recommended by Hadley Wickham, which we follow here.
# But see:
# https://nathaneastwood.github.io/2019/08/18/no-visible-binding-for-global-variable/
# for alternative solutions.
utils::globalVariables("%>%")
utils::globalVariables(c(
  "year",
  "coc_number",
  "unsheltered_chronically_homeless_individuals"
))


#' Counts PIT per year for King County.
#' @return data.frame with one row per year.
#' @export
get_king_pit <- function() {
  ## Load fieldmap, which maps out the column index of all fields
  fieldmap <- rio::import(system.file("extdata", "pit_counts_coc_fieldmap_2007t2018.xlsx", package = "craggy2019"))

  file <- system.file("extdata", "2007-2018-PIT-Counts-by-CoC.xlsx", package = "craggy2019")

  ## Import each year, subset based on the field map, rename columns, and add a `year` field
  y2018 <- rio::import(file, which = 1)[,fieldmap$y2018]
  names(y2018) <- fieldmap$shortname
  y2018$year <- 2018

  y2017 <- rio::import(file, which = 2)[,fieldmap$y2017]
  names(y2017) <- fieldmap$shortname
  y2017$year <- 2017

  y2016 <- rio::import(file, which = 3)[,fieldmap$y2016]
  names(y2016) <- fieldmap$shortname
  y2016$year <- 2016

  y2015 <- rio::import(file, which = 4)[,fieldmap$y2015]
  names(y2015) <- fieldmap$shortname
  y2015$year <- 2015

  y2014 <- rio::import(file, which = 5)[,fieldmap$y2014]
  names(y2014) <- fieldmap$shortname
  y2014$year <- 2014

  y2013 <- rio::import(file, which = 6)[,fieldmap$y2013]
  names(y2013) <- fieldmap$shortname
  y2013$year <- 2013

  y2012 <- rio::import(file, which = 7)[,fieldmap$y2012]
  names(y2012) <- fieldmap$shortname
  y2012$year <- 2012

  y2011 <- rio::import(file, which = 8)[,fieldmap$y2011]
  names(y2011) <- fieldmap$shortname
  y2011$year <- 2011

  y2010 <- rio::import(file, which = 9)[,fieldmap$y2010]
  names(y2010) <- fieldmap$shortname
  y2010$year <- 2010

  y2009 <- rio::import(file, which = 10)[,fieldmap$y2009]
  names(y2009) <- fieldmap$shortname
  y2009$year <- 2009

  y2008 <- rio::import(file, which = 11)[,fieldmap$y2008]
  names(y2008) <- fieldmap$shortname
  y2008$year <- 2008

  y2007 <- rio::import(file, which = 12)[,fieldmap$y2007]
  names(y2007) <- fieldmap$shortname
  y2007$year <- 2007


  ## Row-bind all years together
  pit_counts <- rbind(y2018, y2017, y2016, y2015, y2014, y2013, y2012, y2011, y2010, y2009, y2008, y2007) %>%
    dplyr::select(year, coc_number:unsheltered_chronically_homeless_individuals)

  ## Get PIT for King County
  pit_king <- pit_counts %>%
    dplyr::filter(coc_number == "WA-500")

  return(pit_king)
}
